src/Development/IDE/Core/Rules.hs

```diff
@@ -185,11 +186,21 @@ getHomeHieFile f = do
       wait <- lift $ delayedAction $ mkDelayedAction "OutOfDateHie" L.Info $ do
         hsc <- hscEnv <$> use_ GhcSession f
         pm <- use_ GetParsedModule f
-        typeCheckRuleDefinition hsc pm DoGenerateInterfaceFiles
+        source <- getSourceFileSource f
+        typeCheckRuleDefinition hsc pm DoGenerateInterfaceFiles (Just source)
       _ <- MaybeT $ liftIO $ timeout 1 wait
       ncu <- mkUpdater
       liftIO $ loadHieFile ncu hie_f
 
+getSourceFileSource :: NormalizedFilePath -> Action BS.ByteString
+getSourceFileSource nfp = do
+    (_, msource) <- getFileContents nfp
+    case msource of
+        Nothing -> do
+            source <- readFile' (fromNormalizedFilePath nfp)
+            let bsSource = T.encodeUtf8 $ T.pack source
```

Comment:
This seems like a very roundabout way of reading a file: String -> Text -> BS. 
Why not use `BS.readFile` like GHC does? 

https://gitlab.haskell.org/ghc/ghc/-/blob/master/compiler/GHC/Iface/Ext/Ast.hs#L284

Reply:
Good point, thank you! 

Side question, is there some invariant that this filepath exists in the filesystem? Do we need more exception handling here?

Reply:
Hmm, I think existing error handling will deal with this if the file does not exist.